### PR TITLE
Parity for ION Tools

### PR DESCRIPTION
**Pull request by decentralgabe**
full parity with https://github.com/decentralized-identity/ion-tools

### CODE REVIEW

**Comment by decentralgabe** [comment]
# test
copilot:poem

# other test
copilot:summary

**Comment by michaelneale** [comment]
very nice

File: did/ion/operations.go [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+	longFormDID, err := CreateLongFormDID(*recoveryPubKeyJWK, *updatePubKeyJWK, doc)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "generating long form DID")
+	}
+	shortFormDID, err := GetShortFormDIDFromLongFormDID(longFormDID)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "generating short form DID")
+	}
+	suffix := strings.Split(shortFormDID, ":")[2]
```
Comment (andresuribe87) [inline did/ion/operations.go:193]: No need to split.
```go
	suffix := shortFormDID[len("did:ion:"):]
```

But if you really want to be paranoid, then I would recommend `strings.CutPrefix` instead.

File: did/ion/operations.go [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+	resolutionResult, err := did.ParseDIDResolution(body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "resolving did:ion DID<%s>", id)
+	}
+	return resolutionResult, nil
+}
+
+// Anchor submits an anchor operation to the ION node by appending the operations path to the base URL
+// and making a POST request
+func (i Resolver) Anchor(op AnchorOperation) error {
```
Comment (andresuribe87) [inline did/ion/operations.go:97]: A top level package example on how to use this package would be awesome. Something super simple like: here's how you create an ION did: 

```
NewIONDID()
Anchor()
```

File: did/resolver.go [hunk 3]
```diff
@@ -64,3 +65,23 @@ func GetMethodForDID(did string) (Method, error) {
 	}
 	return Method(split[1]), nil
 }
+
+// ParseDIDResolution attempts to parse a DID Resolution Result or a DID Document
+func ParseDIDResolution(resolvedDID []byte) (*DIDResolutionResult, error) {
+	// first try to parse as a DID Resolution Result
+	var result DIDResolutionResult
```
Comment (andresuribe87) [inline did/resolver.go:72]: Note that `DIDResolutionResult` does not have `json` tags in the struct. I had to add them in https://github.com/TBD54566975/ssi-service/blob/4936f10da0bae4b1606f5b2362c32bb6eb02e170/pkg/service/did/resolve/universal.go#L55

File: did/resolver.go [hunk 4]
```diff
@@ -64,3 +65,23 @@ func GetMethodForDID(did string) (Method, error) {
 	}
 	return Method(split[1]), nil
 }
+
+// ParseDIDResolution attempts to parse a DID Resolution Result or a DID Document
+func ParseDIDResolution(resolvedDID []byte) (*DIDResolutionResult, error) {
```
Comment (andresuribe87) [inline did/resolver.go:70]: Mind adding a test for this function?

File: did/web.go [hunk 5]
```diff
@@ -131,67 +131,65 @@ func (d DIDWeb) GetDocURL() (string, error) {
...
+// Resolve fetches and returns the DIDDocument from the expected URL
+// specification: https://w3c-ccg.github.io/did-method-web/#read-resolve
+func (WebResolver) Resolve(did string, _ ResolutionOptions) (*DIDResolutionResult, error) {
+	if !strings.HasPrefix(did, WebPrefix) {
+		return nil, fmt.Errorf("not a did:web DID: %s", did)
+	}
+	didWeb := DIDWeb(did)
+	doc, err := didWeb.Resolve()
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not resolve did:web DID: %s", did)
```
Comment (andresuribe87) [inline did/web.go:192]:
```suggestion
		return nil, errors.Wrapf(err, "resolving did:web DID: %s", did)
```

File: did/resolver.go (hunk 3 above)
Comment (decentralgabe) [reply did/resolver.go:72]: added

File: did/ion/operations.go [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+	}
+	return &Resolver{baseURL: baseURL}, nil
+}
+
+// Resolve resolves a did:ion DID by appending the DID to the base URL with the identifiers path and making a GET request
+func (i Resolver) Resolve(id string, _ did.ResolutionOptions) (*did.DIDResolutionResult, error) {
+	if i.baseURL == "" {
+		return nil, errors.New("resolver URL is empty")
+	}
+	resp, err := http.Get(strings.Join([]string{i.baseURL, "identifiers", id}, "/"))
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply did/ion/operations.go:75]: good idea - done

File: did/resolver.go (hunk 4 above)
Comment (decentralgabe) [reply did/resolver.go:70]: done

File: did/ion/operations.go [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+		return nil, errors.Wrapf(err, "could not resolve with URL: %s", i.baseURL)
+	}
+
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not resolve with response %+v", resp)
+	}
+	if !is2xxStatusCode(resp.StatusCode) {
+		return nil, fmt.Errorf("could not resolve DID: %s", string(body))
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply did/ion/operations.go:86]: nice

File: did/ion/operations.go [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+func (d *DID) Operations() []any {
+	return d.operations
+}
+
+func (d *DID) Operation(index int) any {
+	return d.operations[index]
+}
+
+// NewIONDID creates a new ION DID with a new recovery and update key pair in addition to any content
+// passed into in the document parameter. The document parameter is optional and can be nil.
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply did/ion/operations.go:153]: it shouldn't be nil, updated comment

File: did/ion/operations.go [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+		longFormDID:        longFormDID,
+		operations:         []any{createRequest},
+		updatePrivateKey:   *updatePrivKeyJWK,
+		recoveryPrivateKey: *recoveryPrivKeyJWK,
+	}, createRequest, nil
+}
+
+// Update updates the DID object's state with a provided state change object. The result is a new
+// update key pair and an update operation to be submitted to an anchor service.
+func (d *DID) Update(stateChange StateChange) (*UpdateRequest, error) {
```
[the remark this answers is not in the document]
Comment (decentralgabe) [reply did/ion/operations.go:207]: good point - i had it like this at first and switched, but agree immutable is safer

**Comment by decentralgabe** [comment]
thanks for the thorough review @andresuribe87 ; should have addressed all

**Review by andresuribe87 — APPROVED** [review]

**Pull request merged**